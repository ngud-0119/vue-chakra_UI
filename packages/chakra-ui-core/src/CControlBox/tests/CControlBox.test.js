import { CVisuallyHidden, CControlBox, CBox } from '../..'
import { render, userEvent } from '@/tests/test-utils'

// its because when we load components from `../..`
jest.mock('@/packages/chakra-ui-core/src/CToast/index.js', () => {})

const renderComponent = (props) => {
  const inlineAttrs = (props && props.inlineAttrs) || ''
  const base = {
    components: { CVisuallyHidden, CControlBox, CBox },
    template: `
  <CBox as="label" display="flex" border-color="gray.200" align-items="center" cursor="pointer" for="control-radio">
    <!-- This is the sibling input, it's visually hidden -->
    <CVisuallyHidden as="input" id="control-radio" data-testid="hiddenControl" ${inlineAttrs} />
  
    <!-- This is the control box with a check icon as children -->
    <CControlBox
      data-testid="control"
      size="24px"
      border="2px"
      bg="white"
      rounded="full"
      border-color="inherit"
      :_checked="{
              bg: 'green.500',
              borderColor: 'green.500'
          }"
      :_hover="{ borderColor: 'gray.300' }"
      :_focus="{ boxShadow: 'outline' }"
      :_disabled="{ opacity: '40%' }"
      ${inlineAttrs}
    >
      <CBox w="50%" h="50%" bg="white" rounded="full" />
    </CControlBox>
  </CBox>
  
    `,
    ...props
  }
  return render(base)
}

it('should render correctly', () => {
  const inlineAttrs = `type="radio"`
  const { asFragment } = renderComponent({ inlineAttrs })
  expect(asFragment()).toMatchSnapshot()
})

test('Uncontrolled radio - should be checked always', async () => {
  const inlineAttrs = `type="radio"`
  const { container, getByTestId } = renderComponent({ inlineAttrs })
  const input = container.querySelector('input')
  const control = getByTestId('control')
  const hiddenControl = getByTestId('hiddenControl')

  // click the first time, it's checked
  await userEvent.click(control)
  expect(input).toBeChecked()
  expect(hiddenControl).toBeChecked()

  // click the second time, it should be still checked
  await userEvent.click(control)
  expect(input).toBeChecked()
  expect(hiddenControl).toBeChecked()
})

test('Uncontrolled checkbox - should be checked always', async () => {
  const inlineAttrs = `type="checkbox"`
  const { container, getByTestId } = renderComponent({ inlineAttrs })
  const input = container.querySelector('input')
  const control = getByTestId('control')
  const hiddenControl = getByTestId('hiddenControl')

  // click the first time, it's checked
  await userEvent.click(control)
  expect(input).toBeChecked()
  expect(hiddenControl).toBeChecked()

  // click the second time, it's unchecked
  await userEvent.click(control)
  expect(input).not.toBeChecked()
  expect(hiddenControl).not.toBeChecked()
})
