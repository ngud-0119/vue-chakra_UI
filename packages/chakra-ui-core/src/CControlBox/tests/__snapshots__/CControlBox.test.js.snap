// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-color: var(--chakra-colors-gray-200);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-1 {
  border: 0px;
  clip: rect(0px, 0px, 0px, 0px);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0px;
  overflow: hidden;
  white-space: nowrap;
  position: absolute;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 120ms;
  transition: all 120ms;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  border: var(--chakra-borders-2px);
  background: var(--chakra-colors-white);
  border-radius: var(--chakra-radii-full);
  border-color: inherit;
}

.emotion-2[aria-checked=true],
.emotion-2[data-checked] {
  background: var(--chakra-colors-green-500);
  border-color: var(--chakra-colors-green-500);
}

.emotion-2:hover,
.emotion-2[data-hover] {
  border-color: var(--chakra-colors-gray-300);
}

.emotion-2:focus,
.emotion-2[data-focus] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-2[disabled],
.emotion-2[aria-disabled=true],
.emotion-2[data-disabled] {
  opacity: 0.4;
}

input[type=checkbox]:focus+.emotion-3 {
  box-shadow: var(--chakra-shadows-outline);
}

input[type=checkbox]:hover:not(:disabled):not(:checked)+.emotion-3 {
  border-color: var(--chakra-colors-gray-300);
}

input[type=checkbox]:disabled+.emotion-3 {
  opacity: 0.4;
}

.emotion-3>* {
  opacity: 0;
}

input[type=checkbox]:checked+.emotion-3,
input[type=checkbox][aria-checked=mixed]+.emotion-3 {
  background: var(--chakra-colors-green-500);
  border-color: var(--chakra-colors-green-500);
}

input[type=checkbox]:checked+.emotion-3>*,
input[type=checkbox][aria-checked=mixed]+.emotion-3>* {
  opacity: 1;
}

.emotion-4 {
  width: 50%;
  height: 50%;
  background: var(--chakra-colors-white);
  border-radius: var(--chakra-radii-full);
}

<label
    class="emotion-0"
    data-chakra-component="CBox"
    for="control-radio"
  >
    <input
      class="emotion-1"
      data-chakra-component="CVisuallyHidden"
      data-testid="hiddenControl"
      id="control-radio"
      type="radio"
    />
     
    <div
      aria-hidden="true"
      class="emotion-2 emotion-3"
      data-chakra-component="CControlBox"
      data-testid="control"
    >
      <div
        class="emotion-4"
        data-chakra-component="CBox"
      />
    </div>
  </label>
</DocumentFragment>
`;
